common/utilities/timing.py

```diff
@@ -0,0 +1,127 @@
+        return diff
+
+
+def time_method(method):
+    """
+    A method decorator that logs the time taken to execute a given method
+
+    :param method:
+    :return:
+    """
+
+    @wraps(method)
+    def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
+                'duration': duration
+            })
+
+    return invoke_method_with_timer
+
+
+def time_method_async(method):
+    """
+    A method decorator that logs the time taken to execute a given method. This decorator should be used to wrap
+    asynchronous methods as the standard time_method decorator does not handle async methods effectively
+    """
+
+    @wraps(method)
+    async def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return await method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
+                'duration': duration
+            })
+
+    return invoke_method_with_timer
+
+
+def log_request(method):
+    """
+    A method to be used with tornado end points to extract their calling details and time their execution, this
+    mainly holds as a placeholder if any extra data is required from the call
+    """
+    def method_wrapper(*args, **kwargs):
+
+        # Gets the 'self' of the caller
+        handler = args[0]
+
+        log_details = {
+            'handler': handler.__class__.__name__,
+            'method': handler.request.method.lower(),
+            'requestBody': handler.request.body
+        }
+
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+
+        try:
+            return method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            log_details['duration'] = duration
+            logging.info(log_details)
+
+    return method_wrapper
+
+
+def async_log_request(method):
+    """
+    A method to be used with tornado end points to extract their calling details and time their execution, this
+    mainly holds as a placeholder if any extra data is required from the call. This decorator should be used for
+    asynchronous tornado calls to ensure the async methods called are handled properly
+    """
+    async def method_wrapper(*args, **kwargs):
+
+        # Gets the 'self' of the caller
+        handler = args[0]
+
+        log_details = {
+            'handler': handler.__class__.__name__,
+            'method': handler.request.method.lower()
+        }
+
+        if handler.request.body:
```

Comment:
I'm not sure we should be logging the full body of requests. I can foresee that these may contain patient data that we don't want to log. Also, if we do keep this, we should presumably have the same `if handler.request.body` block in `log_request` above.

Reply:
I did wonder about this as spine does something similar, however as a part of the logging module particular patient data is hidden but it seems quite difficult, should we just avoid logging anything that might contain any data just as a catch all?

Reply:
I think logging the full request would also violate the "Do not log extremely verbose data" guidance in the "Other Logging Good Practice" section of our [logging strategy](https://gpitbjss.atlassian.net/wiki/spaces/RTDel/pages/1560608910/KDD-001+-+Logging+Strategy+and+Format), so I'm inclined to say we should avoid it. Having this method that extracts key data would seem to meet the spirit of that section. Probably one to confirm with Dave.